File: generateParentheses.swift
class Solution {
    func generateParenthesis(_ n: Int) -> [String] {
        var result = [String]()
        var path = [Character]()
        
        func dfs(_ open: Int, _ close: Int) {
            if path.count == 2 * n {
                result.append(String(path))
                return
            }
            if open < n {
                path.append("(")
                dfs(open + 1, close)
                path.removeLast()
            }
            if close < open {
                path.append(")")
                dfs(open, close + 1)
                path.removeLast()
            }
        }
        
        dfs(0, 0)
        return result
    }
}

File: generateParentheses.go
package main

import "fmt"

func generateParenthesis(n int) []string {
    var res []string
    var dfs func(open, close int, cur []byte)
    dfs = func(open, close int, cur []byte) {
        if len(cur) == 2*n {
            res = append(res, string(cur))
            return
        }
        if open < n {
            dfs(open+1, close, append(cur, '('))
        }
        if close < open {
            dfs(open, close+1, append(cur, ')'))
        }
    }
    dfs(0, 0, make([]byte, 0, 2*n))
    return res
}

func main() {
    fmt.Println(generateParenthesis(3))
}
